secboot/secboot_tpm.go

```diff
@@ -69,3 +95,314 @@ func checkSecureBootEnabled() error {
+			PCRAlgorithm:   tpm2.HashAlgorithmSHA256,
+			PCRIndex:       tpmPCR,
+			KernelCmdlines: params.KernelCmdlines,
+		}
+		if err := sbAddSystemdEFIStubProfile(pcrProfile, &systemdStubParams); err != nil {
+			return fmt.Errorf("cannot add systemd EFI stub profile: %v", err)
+		}
+	}
+
+	// Add snap model profile
+	if len(params.Models) != 0 {
+		snapModelParams := sb.SnapModelProfileParams{
+			PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+			PCRIndex:     tpmPCR,
+			Models:       params.Models,
+		}
+		if err := sbAddSnapModelProfile(pcrProfile, &snapModelParams); err != nil {
+			return fmt.Errorf("cannot add snap model profile: %v", err)
+		}
+	}
+
+	// Provision the TPM as late as possible
+	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		return err
+	}
+	// Seal key to the TPM
+	creationParams := sb.KeyCreationParams{
+		PCRProfile: pcrProfile,
+		PINHandle:  pinHandle,
+	}
+	if err := sbSealKeyToTPM(tpm, key[:], params.KeyFile, params.PolicyUpdateDataFile, &creationParams); err != nil {
+		return fmt.Errorf("cannot seal data: %v", err)
+	}
+
+	return nil
+
+}
+
+func tpmProvision(tpm *sb.TPMConnection, lockoutAuthFile string) error {
+	// Create and save the lockout authorization file
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err := rand.Read(lockoutAuth)
+	if err != nil {
+		return fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+	if err := osutil.AtomicWriteFile(lockoutAuthFile, lockoutAuth, 0600, 0); err != nil {
+		return fmt.Errorf("cannot write the lockout authorization file: %v", err)
+	}
+
+	// TODO:UC20: ideally we should ask the firmware to clear the TPM and then reboot
+	//            if the device has previously been provisioned, see
+	//            https://godoc.org/github.com/snapcore/secboot#RequestTPMClearUsingPPI
+	if err := sbProvisionTPM(tpm, sb.ProvisionModeFull, lockoutAuth); err != nil {
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+	return nil
+}
+
+func buildLoadSequence(filePaths []string) *sb.EFIImageLoadEvent {
+	// The idea of EFIImageLoadEvent is to build a set of load paths for the current
+	// device configuration. So you could have something like this:
+	//
+	// shim -> recovery grub -> recovery kernel 1
+	//                      |-> recovery kernel 2
+	//                      |-> recovery kernel ...
+	//                      |-> normal grub -> run kernel good
+	//                                     |-> run kernel try
+	//
+	// Or it could look like this, which is the same thing:
+	//
+	// shim -> recovery grub -> recovery kernel 1
+	// shim -> recovery grub -> recovery kernel 2
+	// shim -> recovery grub -> recovery kernel ...
+	// shim -> recovery grub -> normal grub -> run kernel good
+	// shim -> recovery grub -> normal grub -> run kernel try
+
+	var event *sb.EFIImageLoadEvent
+	var next []*sb.EFIImageLoadEvent
+
+	for i := len(filePaths) - 1; i >= 0; i-- {
```

Comment:
Why are the files processed in reverse order?

Reply:
We're building the load chain from leaf to root to chain the node created in the previous iteration. It's just for convenience.